### PR TITLE
fix(gemini): map candidate.finish_reason to the OpenAI vocabulary

### LINKED ISSUE #1196: [BUG] Gemini converter discards candidate.finish_reason, so LengthFinishReasonError (#881) never fires for Gemini
**`any-llm` version:** 1.21.0 (`any-llm-sdk`, google-genai 2.13.0, pydantic 2.12.5)
**Python Version**: 3.13

### Describe the bug

#881 introduced `LengthFinishReasonError` / `ContentFilterFinishReasonError` so that users can catch truncated/filtered structured output. That mechanism relies on `choice.finish_reason` being faithful — but the Gemini provider's response converter discards Gemini's real finish reason, so **these exceptions can never fire for Gemini**.

In `providers/gemini/utils.py::_convert_response_to_response_dict` (non-streaming), `finish_reason` is hardcoded — `"tool_calls"` when function calls are present, otherwise `"stop"` for any text content; `candidate.finish_reason` is never read:

```python
        elif text_content:
            choices.append(
                {
                    "message": {...},
                    "finish_reason": "stop",   # <-- Gemini reported MAX_TOKENS; discarded
                    "index": 0,
                }
            )
```

The streaming converter *does* read `candidate.finish_reason`, but only maps `"STOP"`; `MAX_TOKENS`/`SAFETY` fall through to `None`:

```python
    elif candidate.finish_reason and candidate.finish_reason.value == "STOP":
        finish_reason = "stop"
```

Consequence with a pydantic `response_format`: when Gemini hits `max_output_tokens` mid-generation (easy on thinking models — the thinking budget counts against `max_output_tokens`), the truncated text reaches `acompletion()`'s post-processing labeled `finish_reason="stop"`, skips the `== "length"` check, and lands in `parse_json_content()` — which raises a confusing malformed-JSON `ValidationError` instead of `LengthFinishReasonError`:

```
pydantic_core._pydantic_core.ValidationError: 1 validation error for AssessmentResponse
  Invalid JSON: EOF while parsing a string at line 1 column 31
  [type=json_invalid, input_value='{"explanation":"The user\'s goal', input_type=str]
```

We hit this in production: callers that retry `LengthFinishReasonError` with an increased token budget never get the chance for Gemini, and the `ValidationError` misleadingly suggests a model/schema problem rather than truncation.

### Example w/ Code

```python
import asyncio
from pydantic import BaseModel
from any_llm import acompletion

class Assessment(BaseModel):
    explanation: str
    result: bool

async def main():
    # Small budget + thinking model => Gemini returns finishReason=MAX_TOKENS mid-JSON.
    # Expected: LengthFinishReasonError (as for openai). Actual: pydantic ValidationError (json_invalid).
    await acompletion(
        model="gemini:gemini-3-flash-preview",
        messages=[{"role": "user", "content": "Assess: water is wet. Explain briefly."}],
        response_format=Assessment,
        max_tokens=200,
    )

asyncio.run(main())
```

### Suggested fix

Map Gemini's finish reasons in both converters, e.g. `MAX_TOKENS -> "length"` and `SAFETY`/`PROHIBITED_CONTENT` -> `"content_filter"`, falling back to `"stop"` — then the #881 exceptions work uniformly across providers.

Happy to send a PR if the approach sounds right.

### PR DESCRIPTION
**Pull request by yigor**
## Description

The Gemini converters discarded `candidate.finish_reason`: the non-streaming path hardcoded `"tool_calls"`/`"stop"`, and the streaming path only special-cased `STOP`. As a result the #881 exceptions (`LengthFinishReasonError`, `ContentFilterFinishReasonError`) could never fire for Gemini, and truncated structured output surfaced as a misleading malformed-JSON `ValidationError` (see #1196 for a production repro).

Following the approach endorsed in the issue comments:

- **Both converters** now map `candidate.finish_reason` via a shared `_map_finish_reason` helper: `MAX_TOKENS -> "length"`; `SAFETY`, `RECITATION`, `PROHIBITED_CONTENT`, `BLOCKLIST`, `SPII`, `IMAGE_SAFETY`, `IMAGE_PROHIBITED_CONTENT`, `IMAGE_RECITATION` -> `"content_filter"`; `STOP -> "stop"`. Unmapped reasons fall back to `"stop"` for complete responses and stay `None` for stream chunks, so non-terminal chunks are not forced to a terminal reason.
- **Truncation/filtering take priority over `"tool_calls"`** in both paths (shared `_resolve_finish_reason` helper), so a response cut short mid tool call does not look like a completed tool-call round.
- **A choice is emitted even when Gemini returns no content** (e.g. a thinking model spends the whole `max_output_tokens` budget on reasoning), so the #881 guard fires instead of returning empty `choices`. The streaming converter previously `assert`ed on `candidate.content`, so the same empty-content terminal chunk would have crashed there; it now handles it too.
- `vertexai` inherits both converters from `GoogleProvider`, so it picks up the fix automatically.

**Tests:**
- End-to-end regressions through `acompletion` with a pydantic `response_format` asserting `LengthFinishReasonError` and `ContentFilterFinishReasonError` (mocked client, `tests/unit`).
- Converter-level unit tests: exhaustive mapping table for `_map_finish_reason`, representative rows through both converters, the tool-call priority rule, and the empty-content edge cases.
- A real-API integration test (`tests/integration/test_finish_reason.py`) reproducing the issue: small `max_tokens` + pydantic `response_format` must raise `LengthFinishReasonError`. Verified it fails on `main` (malformed-JSON `ValidationError`, 6/6 attempts) and passes with the fix. A live `ContentFilterFinishReasonError` test was deliberately left out: reliably tripping Gemini's safety filter requires keeping prohibited content in the repo and is inherently flaky as Google tunes the filters, so that path is covered at the unit level.

Ran locally: `pre-commit run --all-files` clean, `pytest tests/unit` green (1603 passed), and the full Gemini integration suite with a real key (25 passed, 5 pre-existing "gemini does not support responses" skips).

@njbrake pinging you as offered in the issue, review welcome!

## PR Type

- 🐛 Bug Fix

## Relevant issues

Fixes #1196

## Checklist
- [x] I understand the code I am submitting.
- [x] I have added unit tests that prove my fix/feature works
- [x] I have run this code locally and verified it fixes the issue.
- [x] New and existing tests pass locally
- [x] Documentation was updated where necessary
- [x] I have read and followed the [contribution guidelines](https://github.com/mozilla-ai/any-llm/blob/main/CONTRIBUTING.md)
- [x] **AI Usage:**
    - [ ] No AI was used.
    - [ ] AI was used for drafting/refactoring.
    - [x] This is fully AI-generated.

## AI Usage Information

- AI Model used: Claude Fable 5
- AI Developer Tool used: Claude Code
- Any other info you'd like to share: Implementation, tests, and this PR text were AI-generated under human direction and review; the integration test was validated against the real Gemini API in both directions (fails on main, passes with the fix).

- [x] I am an AI Agent filling out this form (check box if true)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Gemini finish-reason handling for truncation, content filtering, tool calls and normal completion.
  * Preserved response choices when truncated or filtered responses contain no text or tool calls.
  * Structured outputs now raise the appropriate length or content-filter error instead of a parsing error.
  * Improved reliability of streaming responses when terminal content is missing.

* **Tests**
  * Added coverage for Gemini finish-reason mappings, streaming responses and structured-output error handling.

### CODE REVIEW

File: src/any_llm/providers/gemini/utils.py [hunk 1]
```diff
@@ -333,29 +371,19 @@ def _convert_response_to_response_dict(response: types.GenerateContentResponse)
...
-            choices.append(
-                {
-                    "message": {
-                        "role": "assistant",
-                        "content": text_content,
-                        "reasoning": reasoning,
-                        "tool_calls": None,
-                    },
-                    "finish_reason": "stop",
+                    "finish_reason": _resolve_finish_reason(mapped_finish_reason, bool(tool_calls_list)) or "stop",
```
[the remark this answers is not in the document]
Comment (yigor) [reply src/any_llm/providers/gemini/utils.py:386]: Accurate observation, but this is pre-existing behavior rather than something this PR introduced, and I'd treat it as out of scope for #1196:

- On `main`, any candidate without content parts already produced empty `choices`, so `MALFORMED_FUNCTION_CALL` / `UNEXPECTED_TOOL_CALL` responses behaved exactly the same before this change. If anything, this PR narrows the empty-choices window (`length`/`content_filter` now emit a choice); it does not change behavior for those two reasons.
- Nothing crashes downstream: the structured-output post-processing iterates over `choices`, so an empty list returns an empty completion. That is unlike the truncation bug fixed here, where a misleading `ValidationError` was raised.
- Neither `LengthFinishReasonError` nor `ContentFilterFinishReasonError` would be truthful for a malformed or unexpected tool call, and the OpenAI finish-reason vocabulary has no equivalent. Mapping them to `"stop"` with an empty message would arguably be worse: the structured-output path would then silently return `parsed=None` instead of visibly failing. Surfacing these properly likely means a new dedicated exception type, which is a design decision beyond this fix.
- The mapping implemented here is the one the maintainer enumerated in the issue discussion; these reasons were deliberately not part of it.

Happy to open a follow-up issue on what any-llm should surface for `MALFORMED_FUNCTION_CALL` / `UNEXPECTED_TOOL_CALL` if the maintainers think it is worth handling.

**Review by njbrake — APPROVED** [review]
_Note: this review was drafted by Claude via back-and-forth with @njbrake. The reasoning and decisions are his; the prose is Claude's._

Approving. Thanks for the thorough writeup, and for validating the integration test in both directions; that made this much faster to review.

I pushed one commit to your branch (87d18b5): two of the new test lines were over the 120 character limit, so ruff-format (pinned at v0.15.20) rewrapped them, and `run-linter` would have failed. It slipped through because CI on this PR was sitting at `action_required` awaiting approval for a first-time fork contributor, so nothing beyond the template check ever ran. I have approved the workflows and everything is green now.

What I verified beyond reading the diff:

- All ten mapped `FinishReason` members resolve on google-genai 2.14.0. Worth checking, since google-genai is unpinned in pyproject.toml.
- Unknown wire values are safe. google-genai synthesizes a pseudo-member for a reason it does not recognize, so `_FINISH_REASON_MAP.get()` returns `None` and falls through to your `"stop"` / `None` fallbacks rather than raising.
- A differential run of both converters, main versus this branch, over 6 content shapes by 18 finish reasons. Every divergence is intentional: mapped reasons now surface, `length` and `content_filter` emit a choice where main returned `choices == []`, and the streaming path returns a chunk where main raised `AssertionError`. `STOP` with no content still returns `choices == []`, and all six unmapped reasons (`OTHER`, `LANGUAGE`, `NO_IMAGE`, `IMAGE_OTHER`, `MALFORMED_FUNCTION_CALL`, `UNEXPECTED_TOOL_CALL`) behave identically to main. Nothing that previously worked got reclassified.
- Clean merge against current main, with the unit suite and ruff both passing on the merged state. Of the commits main has gained since your merge-base, only #1198 and #1201 touch `any_llm.py`, and neither goes near the finish-reason guard.

I agree with your response to the CodeRabbit finding on `MALFORMED_FUNCTION_CALL` / `UNEXPECTED_TOOL_CALL`. The matrix confirms those two produce identical output on main and on this branch, so it is pre-existing behavior, and a dedicated error contract is the right follow-up rather than folding it into #1196.

I have added the `run-integration-tests` label to get the Gemini suite running here with real keys, since this changes the Gemini response path for all callers and not only structured-output ones.

One follow-up worth doing later, not blocking this: google-genai is unpinned and `_FINISH_REASON_MAP` falls back to `"stop"` for anything new Google adds, so a future safety-flavored reason would quietly regress #1196. A test that iterates `types.FinishReason` and asserts each member is either in the map or in an explicit known-unmapped set would turn that into a loud dependabot failure. It can live entirely in the test file.
